### PR TITLE
Core: Simplify/improve View APIs

### PR DESCRIPTION
**Pull request by nastra**
This PR re-visits the existing View APIs and aims at simplifying/improving a few things, namely:

* `field-aliases` and `field-comments` should be part of the View's schema rather than be part of each individual SQL representation. Both fields are informational, and having them be part of each SQL representation could lead to issues across dialects
* `default-catalog`: should be the name of the current view catalog that loaded this view, thus storing this information isn't required
* `default-namespace`: should be the same across multiple SQL representations. We might want to enforce this and I've added a note in the view spec
* Per dialect there should be no more than one SQL representation. We might want to enforce this and I've added a note in the view spec

## Creating/updating a view with one or more representations

A few options around improving and simplifying the `ViewBuilder` API have been explored. The PR deprecates the existing fine-grained methods on for defining a view's representation and adds a `withQuery(...)` method when creating a View or when updating existing representations:

### Creating a View with one representation
```
View view = catalog.buildView(identifier)
  .withSchema(schema)
  .withQuery("spark", "select a,b from ns.tbl")
  .createOrReplace()
```

### Creating a view with multiple representations
```
View view = catalog.buildView(identifier)
  .withSchema(schema)
  .withQuery("spark", "select a,b from ns.tbl")
  .withQuery("trino", "select a,b from ns.tbl")
  .createOrReplace()
```

### Adding an additional representation for an existing View
```
view.updateRepresentation()
  .withQuery("spark", "select...")  // adds this definition if it doesn't exist or updates the existing one
  .remove("trino")                  // removes the definition for this dialect
  .commit()
```

### Updating a View's properties
```
view.updateProperties()
  .set("key1", "val1")
  .set("key2", "val2")
  .remove("key3")
  .commit()
```

### CODE REVIEW

**Comment by nastra** [comment]
I am actually surprised that RevAPI doesn't fail on those changes. It appears that this is related to to recent Gradle version bump in #7955

File: api/src/main/java/org/apache/iceberg/view/SQLViewRepresentation.java
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/view/SQLViewRepresentation.java:42]: The rationale for removing this isn't necessarily true. We could make this implicit, but it actually causes some problems with how we want to maintain the SQL without modification.

This was added to fill in necessary information that may be missing from the statement that created the view. For example:

```sql
USE prod.default; -- catalog=prod, schema=default
CREATE OR REPLACE VIEW other_catalog.views.new_view (id, row_count) AS
SELECT id, count(1)
FROM some_table
GROUP BY id
```

As you can see, the current catalog and namespace don't necessarily match the catalog and namespace that were current when the view was created. That means we either have to rewrite the SQL text or store the current catalog and namespace from the environment.

File: api/src/main/java/org/apache/iceberg/view/UpdateViewRepresentation.java
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/view/UpdateViewRepresentation.java:32]: I think we need a better name than `UpdateViewRepresentation`. That doesn't capture what is happening.

I'm also not sure we even need this in the initial API. Are these operations that people will need to use? What is the SQL that exercises them or what is the use case?

Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/view/UpdateViewRepresentation.java:32]: One more thing: this may actually be dangerous with multiple people interacting. If you replace a view with a new query at the same time that I add a dialect to the old verison, we could end up mixing SQL that doesn't produce the same result.

**Comment by danielcweeks** [comment]
I feel like we lost some context from #7880, specifically the [discussion](https://github.com/apache/iceberg/pull/7880#discussion_r1250671566) around how we want to represent these APIs.

I agree with @jackye1995 about favoring option B.  I just feel like this is too SQL specific and limits our options going forward with different representations and creates unnecessary differences with the spec representation.

Also, I'm not convinced that you can/should have `field-aliases` and `field-comments` at the top-level.  Some engines may support field aliases, but others do not so they don't apply equally to the representations and is inconsistent with the spec, which has them at the representation level.  

I think it's fine to omit the default catalog and default namespaces, especially if we can appropriately qualify them in the engine when constructing the sql representation.

File: api/src/main/java/org/apache/iceberg/view/View.java
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/view/View.java:92]: I think we want a `newViewVersion()` method instead of this. It should have the same API as the `ViewBuilder`, probably sharing a `ViewVersionBuilder` interface.

File: api/src/main/java/org/apache/iceberg/view/ViewBuilder.java
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/view/ViewBuilder.java:59]: @nastra, what we talked about was that this should be a property of the `ViewVersion`, not the SQL. You can't define two SQL queries with different default namespaces.

File: format/view-spec.md
Comment (rdblue) [inline format/view-spec.md:150]: The view version's schema.

File: format/view-spec.md
Comment (rdblue) [inline format/view-spec.md:104]: Dialect is not something that can be used here because it is defined in the SQL representation. There is no "dialect" for the view version and there could be multiple dialects from multiple SQL representations.

File: api/src/main/java/org/apache/iceberg/view/UpdateViewRepresentation.java
Comment (nastra) [reply api/src/main/java/org/apache/iceberg/view/UpdateViewRepresentation.java:32]: > I'm also not sure we even need this in the initial API. Are these operations that people will need to use? What is the SQL that exercises them or what is the use case?

The main use case would be to have an API to update a view's representation without having to reconstruct all other representations. Say you have representations for `trino` and `spark` and you'd want to only update the SQL definition of `spark`. In a normal `replace` you would replace the existing view completely with the "new" information, but you don't want to lose the representation for `trino`.

File: format/view-spec.md
Comment (nastra) [reply format/view-spec.md:104]: I've removed this sentence as I just noticed that this is already being mentioned:

> A view version can have multiple SQL representations of different dialects, but only one SQL representation per dialect.

File: api/src/main/java/org/apache/iceberg/view/SQLViewRepresentation.java
Comment (nastra) [reply api/src/main/java/org/apache/iceberg/view/SQLViewRepresentation.java:42]: hm good point, should we maybe store `defaultCatalog` at the `ViewVersion` level in this case (same as the `defaultNamespace`?)

File: api/src/main/java/org/apache/iceberg/view/UpdateViewRepresentation.java
Comment (nastra) [reply api/src/main/java/org/apache/iceberg/view/UpdateViewRepresentation.java:32]: So I think the current issue is the semantics of `.replace()` in Iceberg (or maybe I just understood them wrong). 

While a `CREATE OR REPLACE` is clear from an engine's perspective, I don't think the semantics are clear enough from Iceberg's perspective at the Builder level when calling `createOrReplace()`.

## View representations & replace
In #7880 I've implemented `.replace()` as a full replace, meaning that if you had a previous representation for `spark`, it would get completely replaced with whatever new representation was defined (which can be seen [here](url)).
In hindsight I think Iceberg's semantics for `.replace()` should keep previous view representations, so that we get the following behavior: 

```
// create the view when executing "CREATE OR REPLACE VIEW ..." from Spark
View view = catalog.buildView(identifier)
  .withSchema(schema)
  .withQuery("spark", "select a,b from ns.tbl")
  .createOrReplace()

// replace the view when executing "CREATE OR REPLACE VIEW ..." from Trino
view = catalog.buildView(identifier)
  .withSchema(schema)
  .withQuery("trino", "select a,b from ns.tbl")
  .createOrReplace()
```

Checking the representations of the current version of the view would contain both representations
```
// this would contain both representations
assertThat(view.currentVersion().representations()).hasSize(2)
```

## View Properties & replace
To have the same semantics as with view representations during a `replace`, we should keep previous properties, so that `view.properties()` would return `{"key1": "val1", "key2": "val2"}` after executing the below code:

```
// create the view with a property
View view = catalog.buildView(identifier)
  .withSchema(schema)
  .withQuery("spark", "select a,b from ns.tbl")
  .withProperty("key1", "val1")
  .createOrReplace()

// replace the view
view = catalog.buildView(identifier)
  .withSchema(schema)
  .withQuery("trino", "select a,b from ns.tbl")
  .withProperty("key2", "val2")
  .createOrReplace()
```

#7880 currently handles properties during a `replace` as an actual replace and doesn't keep previous properties.

## TLDR
The semantics of `.replace()` at the View API level should mean that previously configured representations and properties are carried over.

That being said, I think if we use these semantics, then we can probably remove the `UpdateViewRepresentation` API (in case we'd ever want to drop a view representation for a particular dialect, then we could add an API that does that)

Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/view/UpdateViewRepresentation.java:32]: > The semantics of .replace() at the View API level should mean that previously configured representations and properties are carried over.

Yes, this is what we do for tables. If you want to erase the history, then you can drop and re-create the object.

> we can probably remove the UpdateViewRepresentation

I think we should remove this. What we want instead is a way to replace the version, like `newVersion()` or `replaceVersion()`. I think that has mostly the same API as the `ViewBuilder`, probably through a shared interface like `ViewVersionBuilder` that has all the version-related methods.

Comment (nastra) [reply api/src/main/java/org/apache/iceberg/view/UpdateViewRepresentation.java]: To summarize: doing a replace should carry over properties, but previous representations will not be carried over (because we don't know whether a previous view representation created for spark is compatible with a new view representation created for trino). I'll add a note to the view spec so that this behavior is clear and doesn't cause surprises.

File: api/src/main/java/org/apache/iceberg/view/ReplaceViewVersion.java
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/view/ReplaceViewVersion.java:40]: I don't think there's a dialect to remove. Everything must be added specifically for this API right?

Comment (nastra) [reply api/src/main/java/org/apache/iceberg/view/ReplaceViewVersion.java:40]: The `.remove()` is mainly to cover this particular use case:
```
View view = catalog.buildView(identifier)
  .withSchema(schema)
  .withQuery("spark", "sql")
  .withQuery("trino", "sql")
  .createOrReplace()

// updates the SQL for trino and removes the representation for spark
view.replaceViewVersion()
  .withQuery("trino", "updated sql")
  .remove("spark")
  .commit()
```

Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/view/ReplaceViewVersion.java:40]: @nastra, I don't think that we want an API that carries over _anything_ from the last version. That's dangerous, so we should remove this method.

Comment (nastra) [reply api/src/main/java/org/apache/iceberg/view/ReplaceViewVersion.java:40]: ok I've removed that method.

File: api/src/main/java/org/apache/iceberg/view/VersionBuilder.java
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/view/VersionBuilder.java:43]: The default namespace should be a property of the view version, not a representation. `withDefaultNamespace` should be a separate method.

File: api/src/main/java/org/apache/iceberg/view/VersionBuilder.java
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/view/VersionBuilder.java:23]: Where are the other methods, like `withSchema` and `withDefaultCatalog`?

File: api/src/main/java/org/apache/iceberg/view/ViewBuilder.java
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/view/ViewBuilder.java:30]: Do we need a `withLocation` to set the location used for storing metadata files?

File: api/src/main/java/org/apache/iceberg/view/VersionBuilder.java
Comment (nastra) [reply api/src/main/java/org/apache/iceberg/view/VersionBuilder.java]: sorry, I was under the impression we would do https://github.com/apache/iceberg/pull/7992#discussion_r1254112125. I've updated the API to properly reflect this and moved the default catalog to the `ViewVersion`

File: api/src/main/java/org/apache/iceberg/view/VersionBuilder.java
Comment (nastra) [reply api/src/main/java/org/apache/iceberg/view/VersionBuilder.java:24]: should be fixed now

File: api/src/main/java/org/apache/iceberg/view/ViewBuilder.java
Comment (nastra) [reply api/src/main/java/org/apache/iceberg/view/ViewBuilder.java:29]: At this point I'm not entirely sure whether we'd need to have a `withLocation()` method here. Should we wait until we get to https://github.com/apache/iceberg/pull/7913 and we can re-evaluate and add it when necessary?

File: api/src/main/java/org/apache/iceberg/view/VersionBuilder.java
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/view/VersionBuilder.java:43]: This Javadoc isn't correct. Looks like a copy/paste problem.

File: api/src/main/java/org/apache/iceberg/view/ViewVersion.java
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/view/ViewVersion.java:84]: Is this optional? What if you don't set it? Does resolution fail at runtime for any unqualified table names?

File: format/view-spec.md
Comment (rdblue) [inline format/view-spec.md:154]: I think "must be" rather than "should be", although this can probably be removed entirely because those are no longer part of the representation. It just IS the same across all representations.

File: .palantir/revapi.yml
Comment (rdblue) [inline .palantir/revapi.yml:763]: I think there's a problem: the generated Immutable classes should not be in the `api` module. All of the implementation classes for API interfaces are in core and it makes no sense to publicly expose these in API when we don't publicly expose the others.

**Review by rdblue — CHANGES_REQUESTED** [review]
I think this is really close to being ready. Overall, I think it's a big improvement to the view definitions and makes behavior more clear and simplifies the API.

The main issue that I think we need to coordinate is the breaking changes in the public API. @aokolnychyi, @jackye1995, @danielcweeks, @Fokko, it would be great to get your input on whether we should move forward with the breaking changes.

In particular, it's **really** unfortunate that I didn't notice that Immutables was generating the implementations in the `api` module until now. That means far more was exposed and will break with these changes -- otherwise we could have simply made the removed methods return `null` or update them to return values from the `ViewVersion` and kept the getters. But the implementations are changing and had public builders as well.

I think since this is an unused API, we can probably get away with the breaking changes but I'd like to insulate us from having this issue in the future by moving the implementation classes into `core` like the others.

File: format/view-spec.md
Comment (nastra) [reply format/view-spec.md:154]: indeed this can be removed now

File: api/src/main/java/org/apache/iceberg/view/VersionBuilder.java
Comment (nastra) [reply api/src/main/java/org/apache/iceberg/view/VersionBuilder.java]: thanks for catching this, fixed

File: api/src/main/java/org/apache/iceberg/view/ViewVersion.java
Comment (nastra) [reply api/src/main/java/org/apache/iceberg/view/ViewVersion.java:84]: Yes this is optional and query resolution would fail at runtime if this isn't set for unqualified names. Engines like Spark or Trino would have to make sure to set this appropriately when the query of the view isn't qualified

File: .palantir/revapi.yml
Comment (nastra) [reply .palantir/revapi.yml:763]: Immutables has a a way to configure the visibility or the package of the generated class, but unfortunately there's nothing currently available that would put the generated class into a different project :(

File: api/src/main/java/org/apache/iceberg/view/ViewVersion.java
Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/view/ViewVersion.java:84]: @nastra, how do you know whether a name is unqualified at runtime?

Comment (nastra) [reply api/src/main/java/org/apache/iceberg/view/ViewVersion.java:84]: I've looked at how Trino handles this. 
Trino always sets the **catalog** and the **schema name** on view creation (even though both fields are optional). During View creation the underlying SQL is parsed and checked. If the SQL is unqualified, it is checked that the table exists within the **catalog/schema** that the view is created under. If the SQL is qualified, it is checked that the qualified table exists.

When a SELECT against a View is performed, Trino parses the [underlying SQL](https://github.com/trinodb/trino/blob/10d3d1f4ef3f9d60581cbb2b5c4e60249abc6dc7/core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java#L2448-L2451) and then sets the [Session scope for the view](https://github.com/trinodb/trino/blob/10d3d1f4ef3f9d60581cbb2b5c4e60249abc6dc7/core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java#L4578) using the **catalog/schema** from the stored view definition. 

The [parsed SQL](https://github.com/trinodb/trino/blob/10d3d1f4ef3f9d60581cbb2b5c4e60249abc6dc7/core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java#L2448) knows whether the table is qualified/unqualified. If it's qualified, then it's executed against the qualified table name. If it's unqualified, then it's executed against the **catalog/schema** that was known at view definition time.

**Comment by nastra** [comment]
> In particular, it's really unfortunate that I didn't notice that Immutables was generating the implementations in the api module until now. That means far more was exposed and will break with these changes -- otherwise we could have simply made the removed methods return null or update them to return values from the ViewVersion and kept the getters. But the implementations are changing and had public builders as well.

It's a mess up on my part and I should have noticed this when `@Value.Immutable` usage was added to `iceberg-api`. I was the main driving factor to get `@Value.Immutable`  into Iceberg and it should have been my responsibility to make sure we're not exposing more than we should, especially in the `iceberg-api` module. 

I have moved the View-specific Immutable classes as part of this PR to `iceberg-core` and I also opened https://github.com/apache/iceberg/pull/8099 to do the same for the other places in `iceberg-api`.

Once both PRs are merged, I'll add a follow-up and mention in the docs that `@Value.Immutable` usage is discouraged on `iceberg-api` (with a checkstyle rule that enforces it) and will also remove the Immutable dependency

File: api/src/main/java/org/apache/iceberg/view/UpdateViewRepresentation.java
Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/view/UpdateViewRepresentation.java]: @nastra, it isn't clear when you talk about carrying over representations because you're missing a layer of metadata. When you replace a view, the previous properties and versions are preserved. Replace also creates an entirely new version of the view that is independent of the other versions and the representations within those versions.

Iceberg is not responsible for somehow merging a view's current version with new SQL provided by the replace.

File: api/src/main/java/org/apache/iceberg/view/ViewVersion.java
Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/view/ViewVersion.java]: I think that this should be required in all cases.

**Comment by rdblue** [comment]
> Once both PRs are merged, I'll add a follow-up and mention in the docs that @Value.Immutable usage is discouraged on iceberg-api (with a checkstyle rule that enforces it) and will also remove the Immutable dependency

I think we should disallow it, not just discourage it. A checkstyle rule sounds like the right approach though, so maybe we're saying the same thing.

File: format/view-spec.md
Comment (rdblue) [inline format/view-spec.md:40]: I think it's incorrect to make this statement in the spec. The "replace" operation for `REPLACE VIEW` is a SQL operation with chosen semantics that are represented in the Iceberg view metadata as you describe. But it's a very different statement to say that "when view metadata is replaced" something happens.

Saying that when metadata is "replaced" here could mean any update to the view metadata object, which is a much broader statement than you intended.

As in the table spec, we don't want to add broad statements like this. You can add a section explaining how a SQL `REPLACE VIEW` works, but even then I wouldn't say that it is a requirement of the spec.

Comment (rdblue) [inline format/view-spec.md:40]:
```suggestion
```

File: format/view-spec.md
Comment (rdblue) [inline format/view-spec.md:41]:
```suggestion
```

File: format/view-spec.md
Comment (rdblue) [inline format/view-spec.md:88]: This should not be optional.

File: format/view-spec.md
Comment (rdblue) [inline format/view-spec.md:90]: This is also incorrect for the spec.

There is nothing preventing us from creating an operation that does exactly this carry over, as long as it is the user's intent. For example:

```sql
ALTER VIEW v ADD REPRESENTATION spark AS
SELECT * FROM foo
```

This SQL clearly **adds** to the existing representations. It isn't correct or incorrect to do that. It is simply the behavior that a SQL operation could have. The spec allows all these things.

Comment (rdblue) [inline format/view-spec.md:90]:
```suggestion
```

File: format/view-spec.md
Comment (rdblue) [inline format/view-spec.md:88]:
```suggestion
| _required_  | `default-namespace` | Namespace to use when a reference in the SELECT is a single identifier        |
```

File: format/view-spec.md
Comment (rdblue) [inline format/view-spec.md:86]:
```suggestion
| _required_  | `default-namespace` | Namespace to use when a reference in the SELECT is a single identifier        |

When `default-catalog` is `null` or not set, the catalog in which the view is stored must be used as the default catalog.
```

**Comment by rdblue** [comment]
I had a few comments on the spec that I added as changes and committed. +1 now, so I'll merge when tests are passing.

**Review by rdblue — APPROVED** [review]

File: format/view-spec.md
Comment (rdblue) [reply format/view-spec.md]: Removed.

File: format/view-spec.md
Comment (rdblue) [reply format/view-spec.md]: Updated to required and added clarification for how to handle this when it is null or missing.

File: format/view-spec.md
Comment (rdblue) [reply format/view-spec.md]: Removed.

**Comment by rdblue** [comment]
Thanks, @nastra!

**Comment by nastra** [comment]
> > Once both PRs are merged, I'll add a follow-up and mention in the docs that @Value.Immutable usage is discouraged on iceberg-api (with a checkstyle rule that enforces it) and will also remove the Immutable dependency
> 
> I think we should disallow it, not just discourage it. A checkstyle rule sounds like the right approach though, so maybe we're saying the same thing.

Yes we are saying the same thing here. The goal with the docs part was mainly to describe why `@Value.Immutable` should not be in `iceberg-api` and show 1-2 examples of how to apply `@Value.Immutable` to an API that is defined in `iceberg-api`.

File: format/view-spec.md
Comment (nastra) [reply format/view-spec.md]: thanks

File: format/view-spec.md
Comment (nastra) [reply format/view-spec.md]: makes sense, thanks for updating this

File: api/src/main/java/org/apache/iceberg/view/ViewVersion.java
Comment (nastra) [reply api/src/main/java/org/apache/iceberg/view/ViewVersion.java]: I've updated this and made it required in https://github.com/apache/iceberg/pull/8156

**Comment by wmoustafa** [comment]
> Also, I'm not convinced that you can/should have `field-aliases` and `field-comments` at the top-level. Some engines may support field aliases, but others do not so they don't apply equally to the representations and is inconsistent with the spec, which has them at the representation level.

+1. Also worth considering whether "Representation" really refers to an engine or a dialect. Right now it looks like dialect, but even if we go with the above we will be making an implicit correlation between dialect and engine.

> I think it's fine to omit the default catalog and default namespaces, especially if we can appropriately qualify them in the engine when constructing the sql representation.

+1. How are we thinking about the fact that catalog names are not the same across engines, even when the underlying tables end up being looked up from the same metastore/catalog? E.g., `hive.db.t` in Trino might refer to the same table as `spark_catalog.db.t` in Spark?

**Pull request merged**
